pkg/indexer/localcredentials.go

```diff
@@ -0,0 +1,83 @@
+package indexer
+
+import (
+	"context"
+	"reflect"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api"
+)
+
+const (
+	AtlasDatabaseUserCredentialsIndex = "atlasdatabaseuser.credentials"
+)
+
+type LocalCredential struct {
+	obj    client.Object
+	name   string
+	logger *zap.SugaredLogger
+}
+
+func NewLocalCredentialsIndexer(name string, obj client.Object, logger *zap.Logger) *LocalCredential {
+	return &LocalCredential{
+		obj:    obj,
+		name:   name,
+		logger: logger.Named(AtlasStreamInstanceByConnectionIndex).Sugar(),
+	}
+}
+
+func (lc *LocalCredential) Object() client.Object {
+	return lc.obj
+}
+
+func (lc *LocalCredential) Name() string {
+	return lc.name
+}
+
+func (lc *LocalCredential) Keys(object client.Object) []string {
+	if reflect.TypeOf(object) != reflect.TypeOf(lc.obj) {
+		lc.logger.Errorf("expected %T but got %T", lc.obj, object)
+		return nil
+	}
+
+	credentialProvider, ok := (object).(api.CredentialsProvider)
+	if !ok {
+		lc.logger.Errorf("expected %T to implement api.CredentialProvider", object)
+		return nil
+	}
+
+	if localRef := credentialProvider.Credentials(); localRef != nil && localRef.Name != "" {
+		return []string{types.NamespacedName{Namespace: object.GetNamespace(), Name: localRef.Name}.String()}
+	}
+	return []string{}
+}
+
+func CredentialsIndexMapperFunc(indexerName string, list api.ReconciliableList, kubeClient client.Client, logger *zap.SugaredLogger) handler.MapFunc {
```

Comment:
is this actually used? 🤔

Reply:
Let me double check whether or not I wired in properly.

Reply:
It was not, but now it is being wired as expected within the database user controller:

```go
func (r *AtlasDatabaseUserReconciler) SetupWithManager(mgr ctrl.Manager, skipNameValidation bool) error {
	return ctrl.NewControllerManagedBy(mgr).
		Named("AtlasDatabaseUser").
		For(&akov2.AtlasDatabaseUser{}, builder.WithPredicates(r.GlobalPredicates...)).
		Watches(
			&corev1.Secret{},
			handler.EnqueueRequestsFromMapFunc(r.findAtlasDatabaseUserForSecret),
			builder.WithPredicates(predicate.ResourceVersionChangedPredicate{}),
		).
		Watches(
			&corev1.Secret{},
			handler.EnqueueRequestsFromMapFunc(indexer.CredentialsIndexMapperFunc(
				indexer.AtlasDatabaseUserCredentialsIndex,
				&akov2.AtlasDatabaseUserList{},
				indexer.DatabaseUserRequests,
				r.Client,
				r.Log,
			)),
			builder.WithPredicates(predicate.ResourceVersionChangedPredicate{}),
		).
		WithOptions(controller.TypedOptions[reconcile.Request]{SkipNameValidation: pointer.MakePtr(skipNameValidation)}).
		Complete(r)
}
```